File: src/main/java/com/schwarzit/spectralIntellijPlugin/models/SpectralIssue.java
package com.schwarzit.spectralIntellijPlugin.models;

import com.intellij.lang.annotation.HighlightSeverity;
import com.intellij.openapi.editor.Document;

public class SpectralIssue {
    private final String code;
    private final String[] path;
    private final String message;
    private final int severity;
    private final String source;
    private final ErrorRange range;

    @SuppressWarnings("unused")
    public SpectralIssue(String code, String[] path, String message, int severity, String source, ErrorRange range) {
        this.code = code;
        this.path = path;
        this.message = message;
        this.severity = severity;
        this.source = source;
        this.range = range;
    }

    public String getCode() {
        return code;
    }

    public String[] getPath() {
        return path;
    }

    public String getMessage() {
        return message;
    }

    public int getSeverity() {
        return severity;
    }

    public String getSource() {
        return source;
    }

    public ErrorRange getRange() {
        return range;
    }

    public static HighlightSeverity mapSeverity(int severity) {
        switch (severity) {
            case 0: // Error
                return HighlightSeverity.ERROR;
            case 1: // Warn
                return HighlightSeverity.WARNING;
            case 2: // Info
                return HighlightSeverity.WEAK_WARNING;
            case 3: // Hint
                return HighlightSeverity.INFORMATION;
            default: {
                return HighlightSeverity.ERROR;
            }
        }
    }

    public void setDocument(Document document) {
        this.range.setDocument(document);
    }
}

File: src/main/kotlin/com/schwarzit/spectralIntellijPlugin/SpectralIssue.kt
package com.schwarzit.spectralIntellijPlugin

import com.intellij.lang.annotation.HighlightSeverity
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement

@Serializable
data class SpectralIssue(
    val code: String,
    // the elements aren't always strings and can be numeric
    val path: List<JsonElement>,
    val message: String,
    val severity: Int,
    val source: String,
    val range: ErrorRange
)


@Serializable
data class ErrorRange(
    val start: ErrorPosition,
    val end: ErrorPosition
)

@Serializable
data class ErrorPosition(
    val line: Int,
    val character: Int
)

fun mapSeverity(severity: Int): HighlightSeverity {
    return when (severity) {
        0 -> HighlightSeverity.ERROR
        1 -> HighlightSeverity.WARNING
        2 -> HighlightSeverity.WEAK_WARNING
        3 -> HighlightSeverity.INFORMATION
        else -> HighlightSeverity.ERROR
    }
}